src/substrait/textplan/Any.h

```diff
@@ -0,0 +1,67 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+
+#pragma once
+
+#include <any>
+
+#include "fmt/format.h"
+
+namespace io::substrait::textplan {
+
+template <typename ValueType>
+inline const ValueType*
+any_cast(const std::any* value, const char* file, int line) noexcept {
+  try {
+    auto result = std::any_cast<ValueType>(value);
+    if (result == nullptr) {
+      throw std::invalid_argument(
+          fmt::format("{}:{} - {}", file, line, "bad any cast"));
+    }
+    return result;
+  } catch (std::bad_any_cast& ex) {
+    throw std::invalid_argument(
+        fmt::format("{}:{} - {}", file, line, "bad any cast"));
+  }
+}
+
+template <typename ValueType>
+inline ValueType*
+any_cast(std::any* value, const char* file, int line) noexcept {
+  try {
+    auto result = std::any_cast<ValueType>(value);
+    if (result == nullptr) {
+      throw std::invalid_argument(
+          fmt::format("{}:{} - {}", file, line, "bad any cast"));
+    }
+    return result;
+  } catch (std::bad_any_cast& ex) {
+    throw std::invalid_argument(
+        fmt::format("{}:{} - {}", file, line, "bad any cast"));
+  }
+}
+
+template <class ValueType>
+inline ValueType any_cast(const std::any& value, const char* file, int line) {
+  try {
+    return std::any_cast<ValueType>(value);
+  } catch (std::bad_any_cast& ex) {
+    throw std::invalid_argument(
+        fmt::format("{}:{} - {}", file, line, "bad any cast"));
+  }
+}
+
+template <class ValueType>
+inline ValueType any_cast(std::any& value, const char* file, int line) {
+  try {
+    return std::any_cast<ValueType>(value);
+  } catch (std::bad_any_cast& ex) {
+    throw std::invalid_argument(
+        fmt::format("{}:{} - {}", file, line, "bad any cast"));
+  }
+}
+
+// A wrapper around std::any_cast that provides exceptions with line numbers.
+#define ANY_CAST(ValueType, Value) \
+  ::io::substrait::textplan::any_cast<ValueType>(Value, __FILE__, __LINE__)
```

Comment:
The problem that this wrapper addresses is the lack of location provided in the exceptions that it does raise (and shouldn't be able to because it's not a macro).  If you get back an invalid any_cast exception without a location (and if you're under a googletest the stack trace will be swallowed unless you know where it's happening) it's really hard to debug.  With this wrapper the output contains the file location which you can click on and be right where the problem occurred.  Without it I believe the steps are:

1. enter the offending test case (probably by debugging just that test)
2. step over the parse and the matcher steps to see if the any_cast exception is there
3. examine the stack trace

Any changes to the ecosystem often yield any_cast errors (I've changed the type being stored for some representations a few times and it is easy to miss some calls).  If the module is going to be maintainable by someone besides the author I believe this wrapper is required.  I agree that it's not a desirable situation which is probably one point in favor of using status objects instead of system level exceptions.

Addendum:  An alternative would be to check the any_cast to see if it contains what is it supposed to contain instead of using a blind any_cast call (allowing the user to catch the issue before causing an exception).  The only disadvantage there is that it will end up making calls more verbose unless we use a macro (or macros) to hide the check/cast.

Reply:
Got it.  So the challenge in this case is that `any_cast` is both potentially exception and very frequent in the code.  Hopefully we can avoid that case going forwards.